api/include/opentelemetry/context/context.h

```diff
@@ -0,0 +1,145 @@
+#pragma once
+
+#include <atomic>
+
+#include "opentelemetry/context/context_value.h"
+#include "opentelemetry/nostd/shared_ptr.h"
+#include "opentelemetry/nostd/string_view.h"
+#include "opentelemetry/trace/key_value_iterable_view.h"
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace context
+{
+
+// The context class provides a context identifier. Is built as a linked list
+// of DataList nodes and each context holds a shared_ptr to a place within
+// the list that determines which keys and values it has access to. All that
+// come before and none that come after.
+class Context
+{
+
+public:
+  // Creates a context object from a map of keys and identifiers, this will
+  // hold a shared_ptr to the head of the DataList linked list
+  template <class T, nostd::enable_if_t<trace::detail::is_key_value_iterable<T>::value> * = nullptr>
+  Context(const T &keys_and_values)
```

Comment:
Will we need the functionality to retrieve a parent context for a context? Will this be needed for modifying the current context at some point?

Reply:
I can't think of an instance, can you?

Reply:
I'm not 100% sure, maybe I'm missing some part of the high-level view.

I think we'll use `Context` to obtain the currently active span: we will have a method to obtain the current context, the current context will hold the `SpanContext` of the active span. I'm trying to figure out necessary steps to make this possible:

1. When a new span starts, we need to set a new current context. We will be able to get this new context via `SetValue("current_span", span_context)`. Then the new span will be stored in the new current context.
2. When the span ends, we will need to restore the previous context, as this should be the current context now (returning the parent span as active span).

I think we'll need some kind of parenting of contexts, but I'm not sure if it will be managed by the `Context` itself or by some other class.

Reply:
I believe that will be handled within the threadlocal context class

Reply:
The implementation of which is described by the context spec

Reply:
Or rather, interface of